plugins/producer_plugin/include/eosio/producer_plugin/production_pause_vote_tracker.hpp

```diff
@@ -0,0 +1,260 @@
+   bool record_received_other_vote(fc::time_point vote_received_time) {
+      return latest_other_vote.record_received_vote(vote_received_time);
+   }
+
+   // Can be called concurrently with all member functions.
+   //
+   // Returns true if vote was recorded.
+   //
+   // The only reason a producer vote would not be recorded is if its received time is
+   // not more recent than the received time of a previously recorded producer vote.
+   bool record_received_producer_vote(fc::time_point vote_received_time) {
+      return latest_producer_vote.record_received_vote(vote_received_time);
+   }
+
+   // Can be called concurrently with all member functions except:
+   //    + set_block_acceptance_tolerance
+   //
+   // This function only records the received block if it is relevant for the correct functioning of
+   // the check_pause_status function evaluated at times no older than the received time of the block.
+   //
+   // Returns true if block was recorded.
+   //
+   // A block will only be recorded if all the following conditions hold for either
+   // the latest recorded other vote or the latest recorded producer vote:
+   //   1. The received time of the block must be more recent than received time of latest recorded
+   //      (other/producer) vote by more than the current block acceptance threshold.
+   //   2. If there is a still tracked recorded block with a received time more recent than the received time of
+   //      the latest recorded (other/producer) vote, then that block's received time must be more recent than
+   //      the received time of the new block to record.
+   bool record_received_block(fc::time_point block_received_time) {
+      bool recorded = false;
+      recorded |= latest_other_vote.record_received_block(block_received_time, block_acceptance_delta);
+      recorded |= latest_producer_vote.record_received_block(block_received_time, block_acceptance_delta);
+      return recorded;
+   }
+
+   // Can be called concurrently with all member functions.
+   //
+   // This function loses track of any previously recorded blocks.
+   void force_unpause() {
+      latest_other_vote.reset_received_blocks();
+      latest_producer_vote.reset_received_blocks();
+   }
+
+   struct pause_status {
+      struct received_times {
+         fc::time_point                latest_vote{};
+         std::optional<fc::time_point> earliest_conflict_block{}; // If present, must be greater than latest_vote.
+      };
+
+      enum class pause_reason : unsigned int {
+         not_paused        = 0b00,
+         old_other_vote    = 0b01,
+         old_producer_vote = 0b10,
+         old_votes         = old_other_vote | old_producer_vote,
+      };
+
+      received_times other_vote{};
+      received_times producer_vote{};
+
+      pause_reason reason() const {
+         unsigned int r = static_cast<unsigned int>(pause_reason::not_paused);
+
+         if (other_vote.earliest_conflict_block.has_value()) {
```

Comment:
Can those `static_cast` be avoided? You can check if both other_vote and producer_vote have values first and set `r` to `old_votes`, and do the rest of checks. But no need to change.

Reply:
Updated to use if/else instead of `|=`